src/python/pants/backend/java/compile/javac.py

```diff
@@ -260,21 +265,41 @@ async def compile_java_source(
             use_nailgun=jdk_setup.digest,
             append_only_caches=jdk_setup.append_only_caches,
             output_directories=(dest_dir,),
-            description=f"Compile {request.component.members} with javac",
+            description=f"Compile {request.component} with javac",
             level=LogLevel.DEBUG,
         ),
     )
-    output: CompiledClassfiles | None = None
-    if process_result.exit_code == 0:
-        stripped_classfiles_digest = await Get(
-            Digest, RemovePrefix(process_result.output_digest, dest_dir)
+    if compile_result.exit_code != 0:
+        return FallibleCompiledClassfiles.from_fallible_process_result(
+            str(request.component),
+            compile_result,
+            None,
         )
-        output = CompiledClassfiles(stripped_classfiles_digest)
+
+    # Jar.
+    # NB: We jar up the outputs in a separate process because the nailgun runner cannot support
+    # invoking via a `bash` wrapper (since the trailing portion of the command is executed by
+    # the nailgun server). We might be able to resolve this in the future via a Javac wrapper shim.
+    output_file = f"{request.component.representative.address.path_safe_spec}.jar"
+    jar_result = await Get(
+        ProcessResult,
+        Process(
+            argv=[
+                bash.path,
+                "-c",
+                " ".join(["cd", dest_dir, ";", zip_binary.path, "-r", f"../{output_file}", "."]),
```

Comment:
Abundance of caution suggests using `shlex.quote` on the paths. But not a blocker.

Reply:
Yea, good call. Most of these are literals, but not the `zip_binary.path`.